mdp.ipynb

```diff
@@ -467,6 +483,102 @@
    "source": [
     "Move the slider above to observe how the utility changes across iterations. It is also possible to move the slider using arrow keys or to jump to the value by directly editing the number with a double click. The **Visualize Button** will automatically animate the slider for you. The **Extra Delay Box** allows you to set time delay in seconds upto one second for each time step."
    ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#POLICY ITERATION\n",
+    "\n",
+    "The `policy_iteration(mdp)` function follows closely from fig. 17.7 pseudocode.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def policy_iteration(mdp):\n",
+    "    \"\"\"Solve an MDP by policy iteration [Figure 17.7]\"\"\"\n",
+    "    U = {s: 0 for s in mdp.states}\n",
+    "    pi = {s: random.choice(mdp.actions(s)) for s in mdp.states}\n",
+    "    while True:\n",
+    "        U = policy_evaluation(pi, U, mdp)\n",
+    "        unchanged = True\n",
+    "        for s in mdp.states:\n",
+    "            a = argmax(mdp.actions(s), key=lambda a: expected_utility(a, s, U, mdp))\n",
+    "            if a != pi[s]:\n",
+    "                pi[s] = a\n",
+    "                unchanged = False\n",
+    "        if unchanged:\n",
+    "            return pi"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "\n",
+    "Given a policy `pi` , the `policy_evaluation` function calculates the utility of each state if `pi` were to be executed. The i<sup>th</sup> iteration specifies the action to be taken in state `s`. We can perform any no. of iterations (`k=` 50 here) to give a reasonably good approximation of utilities."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def policy_evaluation(pi, U, mdp, k=50):\n",
```

Comment:
I agree -- we need to use %psource so that we don't have the same code repeated in two places.